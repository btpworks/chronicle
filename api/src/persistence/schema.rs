// @generated automatically by Diesel CLI.

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    activity (id) {
        id -> Integer,
        name -> Text,
        namespace_id -> Integer,
        domaintype -> Nullable<Text>,
        started -> Nullable<Timestamp>,
        ended -> Nullable<Timestamp>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    activity_attribute (activity_id, typename) {
        activity_id -> Integer,
        typename -> Text,
        value -> Text,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    agent (id) {
        id -> Integer,
        name -> Text,
        namespace_id -> Integer,
        domaintype -> Nullable<Text>,
        current -> Integer,
        identity_id -> Nullable<Integer>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    agent_attribute (agent_id, typename) {
        agent_id -> Integer,
        typename -> Text,
        value -> Text,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    association (agent_id, activity_id, role) {
        agent_id -> Integer,
        activity_id -> Integer,
        role -> Nullable<Text>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    attachment (id) {
        id -> Integer,
        namespace_id -> Integer,
        signature_time -> Timestamp,
        signature -> Text,
        signer_id -> Integer,
        locator -> Nullable<Text>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    delegation (delegate_id, responsible_id, activity_id, role) {
        delegate_id -> Integer,
        responsible_id -> Integer,
        activity_id -> Nullable<Integer>,
        role -> Nullable<Text>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    derivation (activity_id, generated_entity_id, used_entity_id, typ) {
        activity_id -> Nullable<Integer>,
        generated_entity_id -> Integer,
        used_entity_id -> Integer,
        typ -> Nullable<Integer>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    entity (id) {
        id -> Integer,
        name -> Text,
        namespace_id -> Integer,
        domaintype -> Nullable<Text>,
        attachment_id -> Nullable<Integer>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    entity_attribute (entity_id, typename) {
        entity_id -> Integer,
        typename -> Text,
        value -> Text,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    generation (activity_id, generated_entity_id) {
        activity_id -> Integer,
        generated_entity_id -> Integer,
        typ -> Nullable<Text>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    hadattachment (entity_id, attachment_id) {
        entity_id -> Integer,
        attachment_id -> Integer,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    hadidentity (agent_id, identity_id) {
        agent_id -> Integer,
        identity_id -> Integer,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    identity (id) {
        id -> Integer,
        namespace_id -> Integer,
        public_key -> Text,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    ledgersync (correlation_id) {
        correlation_id -> Text,
        offset -> Nullable<Text>,
        sync_time -> Nullable<Timestamp>,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    namespace (id) {
        id -> Integer,
        name -> Text,
        uuid -> Text,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    usage (activity_id, entity_id) {
        activity_id -> Integer,
        entity_id -> Integer,
    }
}

diesel::table! {
    use diesel::sql_types::*;
    use common::prov::*;

    wasinformedby (namespace_id, activity_id, informing_activity_id) {
        namespace_id -> Integer,
        activity_id -> Integer,
        informing_activity_id -> Integer,
    }
}

diesel::joinable!(activity -> namespace (namespace_id));
diesel::joinable!(activity_attribute -> activity (activity_id));
diesel::joinable!(agent -> identity (identity_id));
diesel::joinable!(agent -> namespace (namespace_id));
diesel::joinable!(agent_attribute -> agent (agent_id));
diesel::joinable!(association -> activity (activity_id));
diesel::joinable!(association -> agent (agent_id));
diesel::joinable!(attachment -> identity (signer_id));
diesel::joinable!(attachment -> namespace (namespace_id));
diesel::joinable!(delegation -> activity (activity_id));
diesel::joinable!(derivation -> activity (activity_id));
diesel::joinable!(entity -> attachment (attachment_id));
diesel::joinable!(entity -> namespace (namespace_id));
diesel::joinable!(entity_attribute -> entity (entity_id));
diesel::joinable!(generation -> activity (activity_id));
diesel::joinable!(generation -> entity (generated_entity_id));
diesel::joinable!(hadattachment -> attachment (attachment_id));
diesel::joinable!(hadattachment -> entity (entity_id));
diesel::joinable!(hadidentity -> agent (agent_id));
diesel::joinable!(hadidentity -> identity (identity_id));
diesel::joinable!(identity -> namespace (namespace_id));
diesel::joinable!(usage -> activity (activity_id));
diesel::joinable!(usage -> entity (entity_id));
diesel::joinable!(wasinformedby -> activity (activity_id));
diesel::joinable!(wasinformedby -> namespace (namespace_id));

diesel::allow_tables_to_appear_in_same_query!(
    activity,
    activity_attribute,
    agent,
    agent_attribute,
    association,
    attachment,
    delegation,
    derivation,
    entity,
    entity_attribute,
    generation,
    hadattachment,
    hadidentity,
    identity,
    ledgersync,
    namespace,
    usage,
    wasinformedby,
);
