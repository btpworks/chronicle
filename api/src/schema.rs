table! {
    agent (name) {
        name -> Text,
        namespace -> Text,
        uuid -> Text,
        current -> Integer,
    }
}

table! {
    namespace (name) {
        name -> Text,
        uuid -> Text,
    }
}

joinable!(agent -> namespace (namespace));

allow_tables_to_appear_in_same_query!(
    agent,
    namespace,
);
