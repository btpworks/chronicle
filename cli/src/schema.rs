// @generated automatically by Diesel CLI.

diesel::table! {
    namespace (name) {
        name -> Varchar,
    }
}
