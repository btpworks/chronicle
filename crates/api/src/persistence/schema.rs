// @generated automatically by Diesel CLI.

diesel::table! {
    activity (id) {
        id -> Int4,
        external_id -> Text,
        namespace_id -> Int4,
        domaintype -> Nullable<Text>,
        started -> Nullable<Timestamp>,
        ended -> Nullable<Timestamp>,
    }
}

diesel::table! {
    activity_attribute (activity_id, typename) {
        activity_id -> Int4,
        typename -> Text,
        value -> Text,
    }
}

diesel::table! {
    agent (id) {
        id -> Int4,
        external_id -> Text,
        namespace_id -> Int4,
        domaintype -> Nullable<Text>,
        current -> Int4,
        identity_id -> Nullable<Int4>,
    }
}

diesel::table! {
    agent_attribute (agent_id, typename) {
        agent_id -> Int4,
        typename -> Text,
        value -> Text,
    }
}

diesel::table! {
    association (agent_id, activity_id, role) {
        agent_id -> Int4,
        activity_id -> Int4,
        role -> Text,
    }
}

diesel::table! {
    attachment (id) {
        id -> Int4,
        namespace_id -> Int4,
        signature_time -> Timestamp,
        signature -> Text,
        signer_id -> Int4,
        locator -> Nullable<Text>,
    }
}

diesel::table! {
    delegation (responsible_id, delegate_id, activity_id, role) {
        delegate_id -> Int4,
        responsible_id -> Int4,
        activity_id -> Int4,
        role -> Text,
    }
}

diesel::table! {
    derivation (activity_id, used_entity_id, generated_entity_id, typ) {
        activity_id -> Int4,
        generated_entity_id -> Int4,
        used_entity_id -> Int4,
        typ -> Int4,
    }
}

diesel::table! {
    entity (id) {
        id -> Int4,
        external_id -> Text,
        namespace_id -> Int4,
        domaintype -> Nullable<Text>,
        attachment_id -> Nullable<Int4>,
    }
}

diesel::table! {
    entity_attribute (entity_id, typename) {
        entity_id -> Int4,
        typename -> Text,
        value -> Text,
    }
}

diesel::table! {
    generation (activity_id, generated_entity_id) {
        activity_id -> Int4,
        generated_entity_id -> Int4,
        typ -> Nullable<Text>,
    }
}

diesel::table! {
    hadattachment (entity_id, attachment_id) {
        entity_id -> Int4,
        attachment_id -> Int4,
    }
}

diesel::table! {
    hadidentity (agent_id, identity_id) {
        agent_id -> Int4,
        identity_id -> Int4,
    }
}

diesel::table! {
    identity (id) {
        id -> Int4,
        namespace_id -> Int4,
        public_key -> Text,
    }
}

diesel::table! {
    ledgersync (tx_id) {
        tx_id -> Text,
        bc_offset -> Nullable<Text>,
        sync_time -> Nullable<Timestamp>,
    }
}

diesel::table! {
    namespace (id) {
        id -> Int4,
        external_id -> Text,
        uuid -> Text,
    }
}

diesel::table! {
    usage (activity_id, entity_id) {
        activity_id -> Int4,
        entity_id -> Int4,
    }
}

diesel::table! {
    wasinformedby (activity_id, informing_activity_id) {
        activity_id -> Int4,
        informing_activity_id -> Int4,
    }
}

diesel::joinable!(activity -> namespace (namespace_id));
diesel::joinable!(activity_attribute -> activity (activity_id));
diesel::joinable!(agent -> identity (identity_id));
diesel::joinable!(agent -> namespace (namespace_id));
diesel::joinable!(agent_attribute -> agent (agent_id));
diesel::joinable!(association -> activity (activity_id));
diesel::joinable!(association -> agent (agent_id));
diesel::joinable!(attachment -> identity (signer_id));
diesel::joinable!(attachment -> namespace (namespace_id));
diesel::joinable!(delegation -> activity (activity_id));
diesel::joinable!(derivation -> activity (activity_id));
diesel::joinable!(entity -> attachment (attachment_id));
diesel::joinable!(entity -> namespace (namespace_id));
diesel::joinable!(entity_attribute -> entity (entity_id));
diesel::joinable!(generation -> activity (activity_id));
diesel::joinable!(generation -> entity (generated_entity_id));
diesel::joinable!(hadattachment -> attachment (attachment_id));
diesel::joinable!(hadattachment -> entity (entity_id));
diesel::joinable!(hadidentity -> agent (agent_id));
diesel::joinable!(hadidentity -> identity (identity_id));
diesel::joinable!(identity -> namespace (namespace_id));
diesel::joinable!(usage -> activity (activity_id));
diesel::joinable!(usage -> entity (entity_id));

diesel::allow_tables_to_appear_in_same_query!(
    activity,
    activity_attribute,
    agent,
    agent_attribute,
    association,
    attachment,
    delegation,
    derivation,
    entity,
    entity_attribute,
    generation,
    hadattachment,
    hadidentity,
    identity,
    ledgersync,
    namespace,
    usage,
    wasinformedby,
);
