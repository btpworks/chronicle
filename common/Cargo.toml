[package]
build   = "build.rs"
edition = "2021"
name    = "common"
version = "0.1.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
async-graphql = { features = [
  "opentelemetry",
  "chrono",
  "unblock",
  "default",
  "uuid",
], version = "3.0.38" }
async-trait = "0.1.53"
chrono = { version = "0.4.19", features = ["serde"] }
custom_error = "1.9.2"
derivative = "2.2.0"
diesel = { version = "2.0.0-rc.0", features = [
  "sqlite",
  "uuid",
  "chrono",
  "r2d2",
] }
futures = "0.3.21"
glob = "0.3.0"
hex = "0.4.3"
iref = "2.1.2"
iref-enum = "2.0.0"
json = "0.12.4"
json-ld = "0.4.0"
k256 = { version = "0.11.3", features = [
  "default",
  "arithmetic",
  "ecdsa",
  "pkcs8",
  "sha256",
  "std",
  "pem",
] }
lazy_static = "1.4.0"
openssl      = "0.10.38"
percent-encoding = "2.1.0"
pkcs8 = { version = "*", features = ["std", "alloc"] }
rand = { version = "0.8.5", features = ["getrandom"] }
rand_core = "0.6.3"
serde = "1.0.136"
serde_derive = "1.0.136"
serde_json = "1.0.79"
static-iref = "2.0.0"
tokio = { version = "1.17.0", features = ["time", "macros", "rt-multi-thread"] }
tracing = "0.1.32"
url = { version = "2.2.2", features = ["serde"] }
uuid = { version = "0.8.2", features = ["serde", "v4"] }

[build-dependencies]
json        = "0.12.4"
lazy_static = "1.4.0"

[dev-dependencies]
insta = { version = "1.14.0", features = ["redactions", "toml"] }
proptest = "1.0.0"

[features]
strict = []
