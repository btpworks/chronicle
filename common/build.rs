use std::{env, fs, path::PathBuf};

include!("./src/context.rs");

fn main() {
    let out_str = env::var("OUT_DIR").unwrap();
    let out_path = PathBuf::from(&out_str);
    let mut out_path = out_path
        .ancestors() // .../target/<debug|release>/build/example-<SHA>/out
        .skip(3) // .../target/<debug|release>
        .next()
        .unwrap()
        .to_owned();
    out_path.push("assets");

    if !out_path.exists() {
        fs::create_dir(&out_path).expect("Could not create assets dir");
    }

    let context = &PROV;

    std::fs::write(
        std::path::Path::new(&format!(
            "{}/context.json",
            out_path.as_os_str().to_string_lossy(),
        )),
        context.pretty(2),
    )
    .unwrap();
}
