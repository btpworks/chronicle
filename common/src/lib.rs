#[macro_use]
extern crate serde_derive;

pub mod ledger;
pub mod models;
pub mod signing;
