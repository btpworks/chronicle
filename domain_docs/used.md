# `prov:Used`

> Usage is the beginning of utilizing an entity by an activity. Before usage,
> the activity had not begun to utilize this entity and could not have been
> affected by the entity.
