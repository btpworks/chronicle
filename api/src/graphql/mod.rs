use std::{convert::Infallible, net::SocketAddr, sync::Arc, time::Duration};

use async_graphql::{
    extensions::Tracing,
    http::{playground_source, GraphQLPlaygroundConfig},
    Context, Error, ErrorExtensions, Object, Schema, Subscription, Upload,
};
use async_graphql_extension_apollo_tracing::{ApolloTracing, ApolloTracingDataExt, HTTPMethod};
use async_graphql_warp::{graphql_subscription, GraphQLBadRequest, GraphQLResponse};
use chrono::{DateTime, NaiveDateTime, Utc};
use common::{
    commands::{
        ActivityCommand, AgentCommand, ApiCommand, ApiResponse, EntityCommand, KeyRegistration,
        PathOrFile,
    },
    prov::{ActivityId, AgentId, EntityId},
};
use custom_error::custom_error;
use derivative::*;
use diesel::{prelude::*, r2d2::Pool};
use diesel::{r2d2::ConnectionManager, Queryable, SqliteConnection};
use futures::Stream;
use tracing::{debug, instrument};
use warp::{
    hyper::{Response, StatusCode},
    Filter, Rejection,
};

use crate::ApiDispatch;

#[derive(Default, Queryable)]
pub struct Agent {
    pub id: i32,
    pub name: String,
    pub namespace: String,
    pub domaintype: Option<String>,
    pub publickey: Option<String>,
    pub current: i32,
}

#[derive(Default, Queryable)]
pub struct Activity {
    pub id: i32,
    pub name: String,
    pub namespace: String,
    pub domaintype: Option<String>,
    pub started: Option<NaiveDateTime>,
    pub ended: Option<NaiveDateTime>,
}

#[derive(Default, Queryable)]
pub struct Entity {
    id: i32,
    name: String,
    namespace: String,
    domaintype: Option<String>,
    signature_time: Option<NaiveDateTime>,
    signature: Option<String>,
    locator: Option<String>,
}

#[Object]
impl Agent {
    async fn namespace(&self) -> &str {
        &self.namespace
    }

    async fn name(&self) -> &str {
        &self.name
    }

    async fn public_key(&self) -> Option<&str> {
        self.publickey.as_deref()
    }

    #[graphql(name = "type")]
    async fn typ(&self) -> &str {
        if let Some(ref typ) = self.domaintype {
            typ
        } else {
            "agent"
        }
    }
}

#[Object]
impl Activity {
    async fn namespace(&self) -> &str {
        &self.namespace
    }

    async fn name(&self) -> &str {
        &self.name
    }

    async fn started(&self) -> Option<DateTime<Utc>> {
        self.started.map(|x| DateTime::from_utc(x, Utc))
    }

    async fn ended(&self) -> Option<DateTime<Utc>> {
        self.ended.map(|x| DateTime::from_utc(x, Utc))
    }

    #[graphql(name = "type")]
    async fn typ(&self) -> &str {
        if let Some(ref typ) = self.domaintype {
            typ
        } else {
            "activity"
        }
    }

    async fn was_associated_with<'a>(
        &self,
        ctx: &Context<'a>,
    ) -> async_graphql::Result<Vec<Agent>> {
        use crate::persistence::schema::wasassociatedwith::{self, dsl};

        let store = ctx.data_unchecked::<Store>();

        let mut connection = store.pool.get()?;

        let res = wasassociatedwith::table
            .filter(dsl::activity.eq(self.id))
            .inner_join(crate::persistence::schema::agent::table)
            .load::<((i32, i32), Agent)>(&mut connection)?;

        Ok(res.into_iter().map(|(_, x)| x).collect())
    }

    async fn used<'a>(&self, ctx: &Context<'a>) -> async_graphql::Result<Vec<Entity>> {
        use crate::persistence::schema::used::{self, dsl};

        let store = ctx.data_unchecked::<Store>();

        let mut connection = store.pool.get()?;

        let res = used::table
            .filter(dsl::activity.eq(self.id))
            .inner_join(crate::persistence::schema::entity::table)
            .load::<((i32, i32), Entity)>(&mut connection)?;

        Ok(res.into_iter().map(|(_, x)| x).collect())
    }
}

#[Object]
impl Entity {
    async fn namespace(&self) -> &str {
        &self.namespace
    }

    async fn name(&self) -> &str {
        &self.name
    }

    #[graphql(name = "type")]
    async fn typ(&self) -> &str {
        if let Some(ref typ) = self.domaintype {
            typ
        } else {
            "entity"
        }
    }

    async fn signature_time(&self) -> Option<DateTime<Utc>> {
        self.signature_time.map(|x| DateTime::from_utc(x, Utc))
    }

    async fn signature(&self) -> Option<&str> {
        self.signature.as_deref()
    }

    async fn locator(&self) -> Option<&str> {
        self.locator.as_deref()
    }

    async fn was_attributed_to<'a>(
        &self,
        ctx: &Context<'a>,
    ) -> async_graphql::Result<Vec<Activity>> {
        use crate::persistence::schema::wasgeneratedby::{self, dsl};

        let store = ctx.data_unchecked::<Store>();

        let mut connection = store.pool.get()?;

        let res = wasgeneratedby::table
            .filter(dsl::entity.eq(self.id))
            .inner_join(crate::persistence::schema::activity::table)
            .load::<((i32, i32), Activity)>(&mut connection)?;

        Ok(res.into_iter().map(|(_, x)| x).collect())
    }

    async fn was_generated_by<'a>(
        &self,
        ctx: &Context<'a>,
    ) -> async_graphql::Result<Vec<Activity>> {
        use crate::persistence::schema::wasgeneratedby::{self, dsl};

        let store = ctx.data_unchecked::<Store>();

        let mut connection = store.pool.get()?;

        let res = wasgeneratedby::table
            .filter(dsl::entity.eq(self.id))
            .inner_join(crate::persistence::schema::activity::table)
            .load::<((i32, i32), Activity)>(&mut connection)?;

        Ok(res.into_iter().map(|(_, x)| x).collect())
    }
}

custom_error! {pub GraphQlError
    Db{source: diesel::result::Error}                           = "Database operation failed",
    DbConnection{source: diesel::ConnectionError}               = "Database connection failed",
    Api{source: crate::ApiError}                                = "API",
}

impl ErrorExtensions for GraphQlError {
    // lets define our base extensions
    fn extend(&self) -> Error {
        Error::new(format!("{}", self)).extend_with(|_err, _e| ())
    }
}

#[derive(Derivative)]
#[derivative(Debug)]
pub struct Store {
    #[derivative(Debug = "ignore")]
    pub pool: Pool<ConnectionManager<SqliteConnection>>,
}

impl Store {
    pub fn new(pool: Pool<ConnectionManager<SqliteConnection>>) -> Self {
        Store { pool }
    }
}

#[derive(Default)]
pub struct Query;

#[Object]
impl Query {
    async fn agent<'a>(
        &self,
        ctx: &Context<'a>,
        name: String,
        namespace: String,
    ) -> async_graphql::Result<Option<Agent>> {
        use crate::persistence::schema::agent::{self, dsl};

        let store = ctx.data_unchecked::<Store>();

        let mut connection = store.pool.get()?;

        Ok(agent::table
            .filter(dsl::name.eq(name).and(dsl::namespace.eq(namespace)))
            .first::<Agent>(&mut connection)
            .optional()?)
    }

    async fn activities_by_time<'a>(
        &self,
        ctx: &Context<'a>,
        types: Vec<String>,
        from_inclusive: Option<DateTime<Utc>>,
        end_exclusive: Option<DateTime<Utc>>,
    ) -> async_graphql::Result<Vec<Activity>> {
        use crate::persistence::schema::activity;
        let store = ctx.data_unchecked::<Store>();

        let mut connection = store.pool.get()?;
        let mut query = activity::table.into_boxed();

        if let Some(start) = from_inclusive {
            query = query.filter(activity::started.gt(start.naive_utc()));
        }

        if let Some(end) = end_exclusive {
            query = query.filter(activity::started.lt(end.naive_utc()));
        }

        for t in types {
            query = query.or_filter(activity::domaintype.eq(t))
        }

        Ok(query.load::<Activity>(&mut connection)?)
    }
}

struct Mutation;

async fn agent_context<'a>(
    namespace: &str,
    res: ApiResponse,
    ctx: &Context<'a>,
) -> async_graphql::Result<Agent> {
    match res {
        ApiResponse::Prov(id, _) => {
            use crate::persistence::schema::agent::{self, dsl};

            let store = ctx.data_unchecked::<Store>();

            let mut connection = store.pool.get()?;

            Ok(agent::table
                .filter(
                    dsl::name
                        .eq(AgentId::from(id).decompose())
                        .and(dsl::namespace.eq(namespace)),
                )
                .first::<Agent>(&mut connection)?)
        }
        _ => unreachable!(),
    }
}

async fn activity_context<'a>(
    namespace: &str,
    res: ApiResponse,
    ctx: &Context<'a>,
) -> async_graphql::Result<Activity> {
    match res {
        ApiResponse::Prov(id, _) => {
            use crate::persistence::schema::activity::{self, dsl};

            let store = ctx.data_unchecked::<Store>();

            let mut connection = store.pool.get()?;

            Ok(activity::table
                .filter(
                    dsl::name
                        .eq(ActivityId::from(id).decompose())
                        .and(dsl::namespace.eq(namespace)),
                )
                .first::<Activity>(&mut connection)?)
        }
        _ => unreachable!(),
    }
}

async fn entity_context<'a>(
    namespace: &str,
    res: ApiResponse,
    ctx: &Context<'a>,
) -> async_graphql::Result<Entity> {
    match res {
        ApiResponse::Prov(id, _) => {
            use crate::persistence::schema::entity::{self, dsl};

            let store = ctx.data_unchecked::<Store>();

            let mut connection = store.pool.get()?;

            Ok(entity::table
                .filter(
                    dsl::name
                        .eq(EntityId::from(id).decompose())
                        .and(dsl::namespace.eq(namespace)),
                )
                .first::<Entity>(&mut connection)?)
        }
        _ => unreachable!(),
    }
}

#[Object]
impl Mutation {
    pub async fn create_agent<'a>(
        &self,
        ctx: &Context<'a>,
        name: String,
        namespace: Option<String>,
        typ: Option<String>,
    ) -> async_graphql::Result<Agent> {
        let api = ctx.data_unchecked::<ApiDispatch>();

        let namespace = namespace.unwrap_or_else(|| "default".to_owned());

        let res = api
            .dispatch(ApiCommand::Agent(AgentCommand::Create {
                name,
                namespace: namespace.clone(),
                domaintype: typ,
            }))
            .await?;

        agent_context(&namespace, res, ctx).await
    }

    pub async fn create_activity<'a>(
        &self,
        ctx: &Context<'a>,
        name: String,
        namespace: Option<String>,
        typ: Option<String>,
    ) -> async_graphql::Result<Activity> {
        let api = ctx.data_unchecked::<ApiDispatch>();

        let namespace = namespace.unwrap_or_else(|| "default".to_owned());

        let res = api
            .dispatch(ApiCommand::Activity(ActivityCommand::Create {
                name,
                namespace: namespace.clone(),
                domaintype: typ,
            }))
            .await?;

        activity_context(&namespace, res, ctx).await
    }

    pub async fn generate_key<'a>(
        &self,
        ctx: &Context<'a>,
        name: String,
        namespace: Option<String>,
    ) -> async_graphql::Result<Agent> {
        let api = ctx.data_unchecked::<ApiDispatch>();

        let namespace = namespace.unwrap_or_else(|| "default".to_owned());

        let res = api
            .dispatch(ApiCommand::Agent(AgentCommand::RegisterKey {
                name,
                namespace: namespace.clone(),
                registration: KeyRegistration::Generate,
            }))
            .await?;

        agent_context(&namespace, res, ctx).await
    }

    pub async fn start_activity<'a>(
        &self,
        ctx: &Context<'a>,
        name: String,
        namespace: Option<String>,
        agent: String,
        time: Option<DateTime<Utc>>,
    ) -> async_graphql::Result<Activity> {
        let api = ctx.data_unchecked::<ApiDispatch>();

        let namespace = namespace.unwrap_or_else(|| "default".to_owned());

        let res = api
            .dispatch(ApiCommand::Activity(ActivityCommand::Start {
                name,
                namespace: namespace.clone(),
                time,
                agent: Some(agent),
            }))
            .await?;

        activity_context(&namespace, res, ctx).await
    }

    pub async fn end_activity<'a>(
        &self,
        ctx: &Context<'a>,
        name: String,
        namespace: Option<String>,
        agent: String,
        time: Option<DateTime<Utc>>,
    ) -> async_graphql::Result<Activity> {
        let api = ctx.data_unchecked::<ApiDispatch>();

        let namespace = namespace.unwrap_or_else(|| "default".to_owned());

        let res = api
            .dispatch(ApiCommand::Activity(ActivityCommand::End {
                name: Some(name),
                namespace: Some(namespace.clone()),
                time,
                agent: Some(agent),
            }))
            .await?;

        activity_context(&namespace, res, ctx).await
    }

    pub async fn activity_use<'a>(
        &self,
        ctx: &Context<'a>,
        activity: String,
        name: String,
        namespace: Option<String>,
        typ: Option<String>,
    ) -> async_graphql::Result<Entity> {
        let api = ctx.data_unchecked::<ApiDispatch>();

        let namespace = namespace.unwrap_or_else(|| "default".to_owned());

        let res = api
            .dispatch(ApiCommand::Activity(ActivityCommand::Use {
                name,
                namespace: namespace.clone(),
                domaintype: typ,
                activity: Some(activity),
            }))
            .await?;

        entity_context(&namespace, res, ctx).await
    }

    pub async fn activity_generate<'a>(
        &self,
        ctx: &Context<'a>,
        activity: String,
        name: String,
        namespace: Option<String>,
        typ: Option<String>,
    ) -> async_graphql::Result<Entity> {
        let api = ctx.data_unchecked::<ApiDispatch>();

        let namespace = namespace.unwrap_or_else(|| "default".to_owned());

        let res = api
            .dispatch(ApiCommand::Activity(ActivityCommand::Generate {
                name,
                namespace: namespace.clone(),
                domaintype: typ,
                activity: Some(activity),
            }))
            .await?;

        entity_context(&namespace, res, ctx).await
    }

    pub async fn entity_attach<'a>(
        &self,
        ctx: &Context<'a>,
        name: String,
        namespace: Option<String>,
        attachment: Upload,
        on_behalf_of_agent: String,
        locator: String,
    ) -> async_graphql::Result<Entity> {
        let api = ctx.data_unchecked::<ApiDispatch>();

        let namespace = namespace.unwrap_or_else(|| "default".to_owned());

        let res = api
            .dispatch(ApiCommand::Entity(EntityCommand::Attach {
                name,
                namespace: namespace.clone(),
                agent: Some(on_behalf_of_agent),
                file: PathOrFile::File(Arc::new(Box::pin(
                    attachment.value(ctx)?.into_async_read(),
                ))),
                locator: Some(locator),
            }))
            .await?;

        entity_context(&namespace, res, ctx).await
    }
}

pub struct Subscription;

#[Subscription]
impl Subscription {
    async fn interval(&self, #[graphql(default = 1)] n: i32) -> impl Stream<Item = i32> {
        let mut value = 0;
        async_stream::stream! {
            loop {
                tokio::time::sleep(Duration::from_secs(1)).await;
                value += n;
                yield value;
            }
        }
    }
}

#[instrument]
pub async fn serve_graphql(
    pool: Pool<ConnectionManager<SqliteConnection>>,
    api: ApiDispatch,
    address: SocketAddr,
    open: bool,
) {
    let schema = Schema::build(Query, Mutation, Subscription)
        .extension(ApolloTracing::new(
            "authorization_token".into(),
            "https://yourdomain.ltd".into(),
            "your_graph@variant".into(),
            "v1.0.0".into(),
            10,
        ))
        .extension(Tracing)
        .data(Store::new(pool.clone()))
        .data(api)
        .finish();

    async_graphql_extension_apollo_tracing::register::register(
        "authorization_token",
        &schema,
        "my-allocation-id",
        "variant",
        "1.0.0",
        "staging",
    )
    .await
    .ok();

    let graphql_post = async_graphql_warp::graphql(schema.clone()).and_then(
        |(schema, request): (
            Schema<Query, Mutation, Subscription>,
            async_graphql::Request,
        )| async move {
            Ok::<_, Infallible>(GraphQLResponse::from(
                schema
                    .execute(request.data(ApolloTracingDataExt {
                        userid: None,
                        path: Some("/".to_string()),
                        host: None,
                        method: Some(HTTPMethod::POST),
                        secure: Some(false),
                        protocol: Some("HTTP/1.1".to_string()),
                        status_code: Some(200),
                        client_name: None,
                        client_version: None,
                    }))
                    .await,
            ))
        },
    );

    let graphql_playground = warp::path::end().and(warp::get()).map(|| {
        Response::builder()
            .header("content-type", "text/html")
            .body(playground_source(
                GraphQLPlaygroundConfig::new("/").subscription_endpoint("/"),
            ))
    });

    let open_address = address;
    tokio::spawn(async move {
        tokio::time::sleep(Duration::from_millis(200)).await;
        debug!(?open_address, "Open browser at");
        open::that(&format!("http://{}/", open_address)).ok();
    });

    let routes = graphql_subscription(schema)
        .or(graphql_playground)
        .or(graphql_post)
        .recover(|err: Rejection| async move {
            if let Some(GraphQLBadRequest(err)) = err.find() {
                return Ok::<_, Infallible>(warp::reply::with_status(
                    err.to_string(),
                    StatusCode::BAD_REQUEST,
                ));
            }

            Ok(warp::reply::with_status(
                "INTERNAL_SERVER_ERROR".to_string(),
                StatusCode::INTERNAL_SERVER_ERROR,
            ))
        });

    warp::serve(routes).run(address).await;
}
