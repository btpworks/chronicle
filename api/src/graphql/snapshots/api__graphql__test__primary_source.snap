---
source: api/src/graphql/mod.rs
assertion_line: 999
expression: created
---
[data.hadPrimarySource]
context = 'http://blockchaintp.com/chronicle/ns#entity:generated'
